### PR TITLE
Cache resolved/inferred types for block in blockRegistry

### PR DESCRIPTION
**Pull request by BALEHOK**
Maintain a cache of all blocks with resolved types.

### CODE REVIEW

File: src/blocks/registry.ts [hunk 1]
```diff
@@ -18,8 +18,54 @@
...
+    const blocks = this.cached();
+    for (const block of blocks) {
+      blocksMap[block.id] = {
+        block,
+        // eslint-disable-next-line no-await-in-loop
+        type: await getType(block),
+      };
+    }
+
+    return blocksMap;
```
Comment (BALEHOK) [inline src/blocks/registry.ts:45]: Just for the sake of quick implementation, to be reviewed.

File: src/components/documentBuilder/edit/ElementEditor.tsx [hunk 2]
```diff
@@ -65,21 +64,8 @@ const ElementEditor: React.FC<ElementEditorProps> = ({
...
-        blocksMap[block.id] = {
-          block,
-          // eslint-disable-next-line no-await-in-loop
-          type: await getType(block),
-        };
-      }
-
-      return blocksMap;
-    },
+    async () => blockRegistry.allTyped(),
```
Comment (BALEHOK) [inline src/components/documentBuilder/edit/ElementEditor.tsx:68]: This is the consumer.

**Comment by BALEHOK** [comment]
Hey @twschiller @BLoe 
Please take a look at the proposal for the Blocks Map cache. We use it in 2 (for now) places and build on every render of the components.

**Comment by twschiller** [comment]
I think it's reasonable to pre-compute. I was expecting to see the types also cached? The allTyped method recomputes it each time though?

The implementation in the PR has some race conditions and issues with error handling

**Comment by twschiller** [comment]
@BALEHOK could you split out the BlockMap part of this? There's conflicts now that I've merged in the document editor PR

**Review by twschiller — CHANGES_REQUESTED** [review]
The problem with this approach is that there's a cyclical dependency between the cache and the registry (since getType calls into the registry for composite blocks). Therefore depending on the order bricks are registered some may end up with a "null" type

Therefore we have to ensure all the blocks are already registered before computing the types. (Registering a new block needs to invalidate the cache)

I would recommend:
- Calculating it the first time allTyped is called, not on registration
- Invalidating the cache whenever register is called

I can help write this up in 30min or so

File: src/blocks/registry.ts [hunk 3]
```diff
@@ -18,12 +18,55 @@
...
+    return this.blocksMap;
+  }
+
+  register(...items: IBlock[]): void {
+    super.register(...items);
+
+    // Generating promise for the new chunk of items
+    const nextTypesPromise: Promise<void> = Promise.all(
+      items.map(async (item) =>
+        getType(item).then((itemType) => {
```
Comment (twschiller) [inline src/blocks/registry.ts:49]: Use await here instead of chaining

**Comment by twschiller** [comment]
@BALEHOK made tweaks (made it a Map) and adjusted when the values are computed. Please take a look

**Review by twschiller — APPROVED** [review]

**Pull request merged**